### DISCUSSION TITLE
Grouping the audited model's relation

### DISCUSSION THREAD

**Discussion by thevalo** [owen-it/laravel-auditing#966] [Q&A] [answered]
Is it possible to implement the functionality of grouping the changes, like `event_sequence_id` so that when I'll retrieve the `Invoice` audited log, I could easily associate the `InvoiceItems` that are changed as well? The basic idea is for every audit (create, update, etc.) an random ID would be generated and applied, to associate that change, one `Invoice`  and multiple `InvoiceItems` .

Did anyone stumble upon similar case?

**Comment by thevalo** [comment, marked as answer]
After abusing few of my brain cells, I think I found a solution.

For anyone who needs this in the future:
in the `config/audits.php`:
```
    'resolvers' => [
        'ip_address' => OwenIt\Auditing\Resolvers\IpAddressResolver::class,
        'user_agent' => OwenIt\Auditing\Resolvers\UserAgentResolver::class,
        'url'        => OwenIt\Auditing\Resolvers\UrlResolver::class,
        'event_sequence_id' => \App\AuditResolvers\EventSequenceId::class,
    ],
```

And the resolver itself:
```
class EventSequenceId implements Resolver
{
    public static function resolve(Auditable $auditable)
    {
        $idKey = $auditable instanceof Invoice ? $auditable->id : $auditable->invoice->id;
        return substr(sha1($idKey . $auditable->invoice_no), 0, 16);
    }
}
```

Basically, I get the ID (nanoID), of the `Invoice`, use the `invoice_no` as a salt, so we can generate it when this resolver gets called, once for `Invoice`, and subsequently X times for count of `InvoiceItems`. Just as an example, you can use whatever value you want, as long as it guarantees uniqueness on each audit.

Though it is not recommended to use `updated_at`, because if DB hangs just for a second, then the sequence ID would be different, defeating its purpose. Rely on a unique column for that auditable model, that gets changed, actually.

**Comment by parallels999** [comment]
The problem now is that if you have more model like invoice, you have to add many `$auditable instanceof Invoice`

I solved it another way

```php
$myCurrentInvoice; // Invoice model

$myInvoiceItem = new InvoiceItems ();
$myInvoiceItem->sequence_id = $myCurrentInvoice->getKey(); 
$myInvoiceItem->fill($myInvoiceItemData)->save(); // save works on update/create
```

```php
// maybe another trait
// protected $sequence_id = null; must be added to the model, 
// protected $sequence_no = 'invoice_no'; other auxiliar field
// also transformAudit https://laravel-auditing.com/guide/audit-transformation.html
public function transformAudit(array $data): array{
    $idKey = empty($this->sequence_id ?? null) ? $this->getKey() : $this->sequence_id;
    $noAux = empty($this->sequence_no ?? null) ? '' : $this->{$this->sequence_no};
    $data['event_sequence_id'] = return substr(sha1($idKey . ($noAux ?? '')), 0, 16);

    return $data;
}
```

**Reply by thevalo** [reply]
Yes, I actually have 3 other models with the same logic. However, I think a more robust approach would be dynamically detect a parent a relationship, rather than restructuring migrations. The above code was just an example, works fine for `Invoice`. I'll have to test it with other models, but take a look at this:
```php
 public static function resolve(Auditable $auditable)
    {
        $parentRelation = static::getParentRelation($auditable);

        $idKey = $parentRelation ? $auditable->{$parentRelation}->getKey() : $auditable->getKey();

        return substr(sha1($idKey . $auditable->updated_at), 0, 16);
    }

    /**
     * Dynamically detect a parent relationship.
     */
    private static function getParentRelation($auditable)
    {
        $parentRelations = ['admission', 'offer', 'invoice', 'order']; //define the possible parent relations, if it's a child model

        foreach ($parentRelations as $relation) {
            if (method_exists($auditable, $relation) && $auditable->{$relation}()->exists()) {
                return $relation;
            }
        }

        return null;
    }
```

**Reply by parallels999** [reply]
Better, but i don't like it

Just add a property on Items Model with the parent relation name

```php
public static function resolve(Auditable $auditable)
{
  $parentRelation = $auditable->parentRelation ?? null;

  $idKey = ($parentRelation ? $auditable->{$parentRelation}()?->getKey() : null) ?? $auditable->getKey();

  return substr(sha1($idKey . ($auditable->updated_at ?? '')), 0, 16);
}
```

**Reply by thevalo** [reply]
Yes, but that involves modifying all child models as well, which, depending on the amount of those, can be a pain... maybe.

**Reply by parallels999** [reply]
IMO the same time it takes you to add an item to `$parentRelations` array, you could just add to the model

```php
class InvoiceItem extends Model {
{
    public string $parentRelation = 'invoice';
```
Also better performance in my opinion, and better control

**For example, if a model has a relationship with the same name but it is not of type `Items`?**

**Reply by thevalo** [reply]
Technically, it takes less time to populate one array, rather than adding that property to each model; though I must agree on the performance side.

**Reply by thevalo** [reply]
> Better, but i don't like it
> 
> Just add a property on Items Model with the parent relation name
> 
> ```
> public static function resolve(Auditable $auditable)
> {
>   $parentRelation = $auditable->parentRelation ?? null;
> 
>   $idKey = ($parentRelation ? $auditable->{$parentRelation}()?->getKey() : null) ?? $auditable->getKey();
> 
>   return substr(sha1($idKey . ($auditable->updated_at ?? '')), 0, 16);
> }
> ```

Seems like the `updated_at` does not produce the same output, though.
